### PR TITLE
Unbalanced dataset splits

### PR DESCRIPTION
**Pull request by itrushkin**
This PR introduces a data splitting interface with the following implementations:
- [x] **NumPy arrays**
  - [x] Equal split
  - [x] Random split 
  - [x] LogNormal split
  - [x] Dirichlet split
- [x] **PyTorch datasets**
  - [x] Equal split 
    - [x] in-memory
    - [x] on-disk
  - [x] Random split
    - [x] in-memory
    - [x] on-disk
  - [x] LogNormal split
    - [x] in-memory
    - [x] on-disk
  - [x] Dirichlet split
    - [x] in-memory
    - [x] on-disk
- [x] **Interactive API**

### CODE REVIEW

**Comment by itrushkin** [comment]
Should be merged after #104.

File: openfl-tutorials/interactive_api/Director_Pytorch_Kvasir_UNET/envoy_folder/kvasir_shard_descriptor.py [hunk 1]
```diff
@@ -9,29 +9,25 @@
...
 
-class KvasirShardDescriptor(ShardDescriptor):
-    """Shard descriptor class."""
-
-    def __init__(self, data_folder: str = 'kvasir_data',
-                 rank_worldsize: str = '1,1',
-                 enforce_image_hw: str = None) -> None:
-        """Initialize KvasirShardDescriptor."""
-        super().__init__()
+class KvasirDataset(Dataset):
```
Comment (igor-davidyuk) [inline openfl-tutorials/interactive_api/Director_Pytorch_Kvasir_UNET/envoy_folder/kvasir_shard_descriptor.py]: There is no need in pytorch dataset here. I suggest removing all pytorch mentions

File: openfl-tutorials/interactive_api/Director_Pytorch_Kvasir_UNET/envoy_folder/kvasir_shard_descriptor.py [hunk 2]
```diff
@@ -84,6 +73,39 @@ def __len__(self):
...
+
+        dataset = KvasirDataset(data_folder, enforce_image_hw)
+        labels = [label for _, label in dataset]
+        data_splitter = RandomNumPyDataSplitter()
+        # Settings for sharding the dataset
+        self.rank, self.world_size = tuple(int(num) for num in rank_worldsize.split(','))
+
+        # Sharding
+        shard_idx = data_splitter.split(labels, self.world_size)[self.rank]
+        self.shard = Subset(dataset, shard_idx)
```
Comment (igor-davidyuk) [inline openfl-tutorials/interactive_api/Director_Pytorch_Kvasir_UNET/envoy_folder/kvasir_shard_descriptor.py]: this line could be just 
`self.images_names = [self.images_names[i] for i in shard_idx]`

File: openfl/plugins/data_splitters/numpy.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+    def __init__(self, shuffle=True):
+        """Initialize.
+
+        Args:
+            shuffle(bool): Flag determining whether to shuffle the dataset before splitting.
+        """
+        self.shuffle = shuffle
+
+    def split(self, labels, num_collaborators):
```
Comment (igor-davidyuk) [inline openfl/plugins/data_splitters/numpy.py]: seems like the signature is changed in this subclass

**Review by igor-davidyuk — CHANGES_REQUESTED** [review]
Please fix the integration to the shard_descriptor part.

**Review by igor-davidyuk — APPROVED** [review]
Looks good to me!

**Review by aleksandr-mokrov — APPROVED** [review]

**Comment by itrushkin** [comment]
Jenkins please retry a build

**Comment by itrushkin** [comment]
Jenkins please retry a build

**Comment by itrushkin** [comment]
Jenkins please retry a build

**Pull request merged**